cairo_zero/kakarot/precompiles/precompiles_helpers.cairo

```diff
@@ -33,4 +38,76 @@ namespace PrecompilesHelpers {
+            return TRUE;
+        }
+        if (precompile_address == Constants.CAIRO_MESSAGING_PRECOMPILE) {
+            return TRUE;
+        }
+        return FALSE;
+    }
+
+    // @notice Returns whether the caller is whitelisted to call precompiles.
+    // @param caller_address The address of the caller.
+    // @return Whether the caller is whitelisted.
+    func is_caller_whitelisted{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        caller_address: felt
+    ) -> felt {
+        let (res) = Kakarot_authorized_cairo_precompiles_callers.read(caller_address);
+        return res;
+    }
+
+    // @notice Returns whether the call to the precompile is authorized.
+    // @dev A call is authorized if:
+    // a. The precompile requires a whitelist AND the CODE_ADDRESS of the caller is whitelisted
+    // b. The precompile is CAIRO_BATCH_CALL_PRECOMPILE and the precompile address is the same as the message address (NOT a DELEGATECALL / CALLCODE).
+    // @param precompile_address The address of the precompile.
+    // @param caller_code_address The code_address of the precompile caller.
+    // @param caller_address The address of the caller.
+    // @param message_address The address being executed in the current message.
+    // @return Whether the call is authorized.
+    func is_call_authorized{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        precompile_address: felt,
+        caller_code_address: felt,
+        caller_address: felt,
+        message_address: felt,
+    ) -> felt {
+        alloc_locals;
+        let precompile_requires_whitelist = requires_whitelist(precompile_address);
+
+        // Ensure that calls to precompiles that require a whitelist are properly authorized.
+        if (precompile_requires_whitelist == TRUE) {
+            let is_whitelisted = is_caller_whitelisted(caller_code_address);
+            tempvar syscall_ptr = syscall_ptr;
+            tempvar pedersen_ptr = pedersen_ptr;
+            tempvar range_check_ptr = range_check_ptr;
+            tempvar authorized = is_whitelisted;
+        } else {
+            tempvar syscall_ptr = syscall_ptr;
+            tempvar pedersen_ptr = pedersen_ptr;
+            tempvar range_check_ptr = range_check_ptr;
+            tempvar authorized = TRUE;
+        }
+        let syscall_ptr = cast([ap - 4], felt*);
+        let pedersen_ptr = cast([ap - 3], HashBuiltin*);
+        let range_check_ptr = [ap - 2];
+        let authorized = [ap - 1];
+
+        // Ensure that calls to CAIRO_BATCH_CALL_PRECOMPILE are not made through a delegatecall / callcode.
+        if (precompile_address == Constants.CAIRO_BATCH_CALL_PRECOMPILE) {
+            let is_not_delegatecall = Helpers.is_zero(message_address - precompile_address);
+            tempvar authorized = authorized * is_not_delegatecall;
+        } else {
+            tempvar authorized = authorized;
+        }
+        return authorized;
+    }
```

Comment:
```suggestion
    func is_call_authorized{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
        precompile_address: felt,
        caller_code_address: felt,
        caller_address: felt,
        message_address: felt,
    ) -> felt {
        alloc_locals;
        let precompile_requires_whitelist = requires_whitelist(precompile_address);

        // Ensure that calls to precompiles that require a whitelist are properly authorized.
        if (precompile_requires_whitelist == TRUE) {
            return is_caller_whitelisted(caller_code_address);
        }
        
        // Ensure that calls to CAIRO_BATCH_CALL_PRECOMPILE are not made through a delegatecall / callcode.
        return Helpers.is_zero(message_address - precompile_address);
    }
```

Reply:
not sure to follow this comment

> // Ensure that calls to CAIRO_BATCH_CALL_PRECOMPILE are not made through a delegatecall / callcode.

when a solidity contract makes a CALL to 0x75003, then `message_address == precompile_address` but we want to allow only for external calls (`depth = 0`)

Reply:
Why restricting to external calls?

Reply:
This should be a generic Multicall contract that has delegatecall/callcode disabled.

Reply:
I will keep my version as it's 'safer' for very little overhead. there could be in the future a combination of whitelist but can't delegatecall.